src/services/BaseApi.ts

```diff
@@ -0,0 +1,128 @@
+import axios, { AxiosInstance, AxiosRequestConfig, AxiosRequestHeaders, AxiosResponse } from 'axios'
+
+export type PrefectConfig = {
+  baseUrl: string,
+  token?: string,
+}
+
+export type GetApiBaseUrl<T extends PrefectConfig = PrefectConfig> = (config: T) => string
+export type GetApiHeaders<T extends PrefectConfig = PrefectConfig> = (config: T) => AxiosRequestHeaders
+
+export const getPrefectBaseUrl: GetApiBaseUrl = (config) => {
+  return config.baseUrl
+}
+
+export const getPrefectUIHeaders: GetApiHeaders = () => {
+  return {
+    'X-PREFECT-UI': true,
+  }
+}
+
+export const getAuthorizationHeaders: GetApiHeaders = (config) => {
+  const value: AxiosRequestHeaders = {}
+
+  if (config.token) {
+    value.Authorization = `bearer ${config.token}`
+  }
+
+  return value
+}
+
+export class BaseApi<T extends PrefectConfig = PrefectConfig> {
+  protected readonly apiConfig: T
+  protected routePrefix: string | undefined
+  protected getHeaders: GetApiHeaders[]
+  protected getBaseUrl: GetApiBaseUrl
+
+  public constructor(apiConfig: T) {
+    this.apiConfig = apiConfig
+
+    this.getHeaders = [getPrefectUIHeaders, getAuthorizationHeaders]
+    this.getBaseUrl = getPrefectBaseUrl
+  }
+
+  protected composeBaseUrl(): string {
+    return this.getBaseUrl(this.apiConfig)
+  }
+
+  protected composeHeaders(): AxiosRequestHeaders {
+    return this.getHeaders.reduce((headers, getHeaders) => {
+      return {
+        ...headers,
+        ...getHeaders(this.apiConfig),
+      }
+    }, {})
+  }
+
+  protected removeUndefinedPaths(input: string | undefined): input is string {
+    return !!input
+  }
+
+  protected removeLeadingAndTrailingSlashes(input: string): string {
+    return input.replace(/^(\/)|(\/)$/g, '')
+  }
+
+  protected combinePath(route: string | undefined): string {
+    return [this.routePrefix, route]
+      .filter(this.removeUndefinedPaths)
+      .map(this.removeLeadingAndTrailingSlashes)
+      .join('/')
+  }
```

Comment:
This is just a developer convenience around whether the baseUrl ends with a slash or the route starts with a slash yes?

Reply:
correct. I tried to make slashes more consistent in these PRs but moving forward, ideally developers wouldn't have to know whether to include leading/trailing slashes on routes or routePrefix